include/nbl/video/utilities/IUtilities.h

```diff
@@ -729,7 +757,56 @@ class NBL_API IUtilities : public core::IReferenceCounted
         core::smart_refctd_ptr<CScanner> m_scanner;
     };
 
-}
+class NBL_API ImageRegionIterator
+{
+public:
+    ImageRegionIterator(
+        const core::SRange<const asset::IImage::SBufferCopy>& copyRegions,
+        IPhysicalDevice::SQueueFamilyProperties queueFamilyProps,
+        asset::ICPUBuffer const* srcBuffer,
+        asset::E_FORMAT srcImageFormat,
+        video::IGPUImage* const dstImage
+    );
+    
+    // ! Memory you need to allocate to transfer the remaining regions in one submit.
+    // ! WARN: It's okay to use less memory than the return value of this function for your staging memory, in that usual case more than 1 copy regions will be needed to transfer the remaining regions.
+    size_t getMemoryNeededForRemainingRegions() const;
+
+    // ! Gives `regionToCopyNext` based on `availableMemory`
+    // ! memcopies the data from `srcBuffer` to `stagingBuffer`, preparing it for launch and submit to copy to GPU buffer
+    // ! updates `availableMemory` (availableMemory -= consumedMemory)
+    // ! updates `stagingBufferOffset` based on consumed memory and alignment requirements
+    // ! this function may do format conversions when copying from `srcBuffer` to `stagingBuffer` if srcBufferFormat != dstImage->Format passed as constructor parameters
+    bool advanceAndCopyToStagingBuffer(asset::IImage::SBufferCopy& regionToCopyNext, size_t& availableMemory, size_t& stagingBufferOffset, void* stagingBufferPointer);
```

Comment:
given that a staging buffer cannot be >=4GB you can use `uint32_t` for offsets and sizes

Reply:
why is `stagingBufferOffset` even a reference and not a constant argument?

Reply:
because we will need to advance that + make sure at the beginning that it respects alignment requirements which are not straightforward for the user to ensure. see `addToCurrentUploadBufferOffset` in the implementation

Reply:
ok now its clear.

What do you think about using `uint32_t` ?

Reply:
Yea I guess it's better to use less memory when we can.